### PR TITLE
`ParsedFunctorMaterial`: Back from the grave

### PR DESCRIPTION
**Pull request by joshuahansel**
This adds `ParsedFunctorMaterial`, possibly moosekind's greatest achievement in flexibility.

There was a PR for this: https://github.com/idaholab/moose/pull/22276, but it was struck down by the stale-inator. This attempt tries to be more minimal, for simplicity. I prefer that we not necessarily add all of the bells and whistles here, just make sure that it is correct within its current scope. For example, I have not tried to mimic the inheritance structure of `ParsedMaterial`, which supports `DerivativeParsedMaterial`. I'd prefer to refactor only at the point that someone wants to use that. I have not inherited from `ParsedMaterial` and `ParsedMaterialHelper` because those classes have individual parameters for variables, PPs, material properties, and functions. `ParsedFunctorMaterial` will not support regular material properties, only functor material properties, and the others will be retrieved in a single functors parameter.

Refs #24380

### CODE REVIEW

**Comment by joshuahansel** [comment]
I hereby summon the Council: @lindsayad @GiudGiud @loganharbour @dschwen @roystgnr.

@lindsayad I need to find a good way to get x, y, z from the generic spatial argument `auto r` in a functor material property definition. I'm thinking we could have some kind of functor utility function `getPoint(r)` that is overloaded on the different arguments? Similarly, `Function` had `getTime(time_arg)` to get the actual time value, so I've duplicated it, but I think this could go in a common place too. Should we make a `FunctorUtils.h` or something, or put it `FunctorMaterial` or `FunctorInterface` or what?

@dschwen You had duplication issues with the original PR. Is what I have an acceptable structure? The functor case is just a bit simpler. Also, I still need to do the optimization stuff (see `ParsedMaterialHelper<is_ad>::functionsPostParse()` and `ParsedMaterialHelper<is_ad>::functionsOptimize()`). I wasn't sure if it just needed to be the exact same as in `ParsedMaterialHelper` or what. Note that `ParsedFunctorMaterial` does inherit from `FunctionParserUtils` at least.

Other things to consider: what tests do we need to see? If it complicates things otherwise, I'd prefer to restrict the scope to just `Real`/`ADReal` functors (if so, do I need to do some check in the code?).

**Comment by joshuahansel** [comment]
Also note that I welcome commits to the branch. I hereby forbid force-pushing until we're all ready to merge (we can squash if desired at that point).

**Comment by GiudGiud** [comment]
getPoint implemented in the relevant functor arguments structs seem fine to me.

**Comment by lindsayad** [comment]
`getPoint` on the fly will be expensive for finite element spatial arguments (as they currently stand) because it doesn't have information about the location of the quadrature points in physical space *a priori*. You would need to reinit a FE helper with the quadrature rule and element in order to get those

**Comment by lindsayad** [comment]
We could probably require that the finite element spatial argument constructors take the physical point in space; I think we have those available in all the places we form finite element spatial arguments

**Comment by joshuahansel** [comment]
> We could probably require that the finite element spatial argument constructors take the physical point in space; I think we have those available in all the places we form finite element spatial arguments

Works for me. Would this add any significant expense?

**Comment by lindsayad** [comment]
No it would not

**Comment by joshuahansel** [comment]
@lindsayad What do you recommend I do about this?:
```
// TODO: Make this some common utility function: duplicated from Function::getTime
template <bool is_ad>
Real
ParsedFunctorMaterialTempl<is_ad>::getTime(const Moose::StateArg & state) const
```
In other words, can/should I put this in a common place and where?

**Comment by joshuahansel** [comment]
Ok, this is ready for some review now. Once we're happy with the design, I'll add doc.

File: framework/src/functormaterials/ParsedFunctorMaterial.C [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+InputParameters
+ParsedFunctorMaterialTempl<is_ad>::validParams()
+{
+  InputParameters params = FunctorMaterial::validParams();
+  params += FunctionParserUtils<is_ad>::validParams();
+
+  params.addClassDescription(
+      "Computes a functor material from a parsed expression of other functors.");
+
+  params.addRequiredCustomTypeParam<std::string>(
```
Comment (lindsayad) [inline framework/src/functormaterials/ParsedFunctorMaterial.C:25]:
```suggestion
  params.addClassDescription(
      "Computes a functor material from a parsed expression of other functors.");
  params.addRequiredCustomTypeParam<std::string>(
```

File: framework/src/functormaterials/ParsedFunctorMaterial.C [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+
+        // Store the space and time values
+        const auto r_point = r.getPoint();
+        _func_params[_n_functors] = r_point(0);
+#if LIBMESH_DIM > 1
+        _func_params[_n_functors + 1] = r_point(1);
+#endif
+#if LIBMESH_DIM > 2
+        _func_params[_n_functors + 2] = r_point(2);
+#endif
```
Comment (lindsayad) [inline framework/src/functormaterials/ParsedFunctorMaterial.C:85]: I think we can safely discard the use of these macros here. It hinders readability. If we don't configure libMesh with a dimension of 3, I'd be shocked if a lot of things in MOOSE didn't die horribly

File: framework/src/functormaterials/ParsedFunctorMaterial.C [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+  // Collect the symbols corresponding to the _func_params values
+  std::vector<std::string> symbols(_functor_symbols);
+  std::string symbols_str = Moose::stringify(symbols);
+  symbols_str += ",x";
+#if LIBMESH_DIM > 1
+  symbols_str += ",y";
+#endif
+#if LIBMESH_DIM > 2
+  symbols_str += ",z";
+#endif
```
Comment (lindsayad) [inline framework/src/functormaterials/ParsedFunctorMaterial.C:114]: same comment as above

File: framework/src/functormaterials/ParsedFunctorMaterial.C [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+        // Store the space and time values
+        const auto r_point = r.getPoint();
+        _func_params[_n_functors] = r_point(0);
+#if LIBMESH_DIM > 1
+        _func_params[_n_functors + 1] = r_point(1);
+#endif
+#if LIBMESH_DIM > 2
+        _func_params[_n_functors + 2] = r_point(2);
+#endif
+        _func_params[_n_functors + LIBMESH_DIM] = _fe_problem.getTimeFromStateArg(t);
```
Comment (lindsayad) [inline framework/src/functormaterials/ParsedFunctorMaterial.C:86]:
```suggestion
        _func_params[_n_functors + Moose::dim] = _fe_problem.getTimeFromStateArg(t);
```

File: framework/src/functormaterials/ParsedFunctorMaterial.C [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+  if (_parsed_function->Parse(_expression, symbols_str) >= 0)
+    mooseError("The expression\n'",
+               _expression,
+               "'\nwith symbols\n'",
+               symbols_str,
+               "'\ncould not be parsed:\n",
+               _parsed_function->ErrorMsg());
+
+  // Resize the values vector
+  _func_params.resize(_n_functors + LIBMESH_DIM + 1);
```
Comment (lindsayad) [inline framework/src/functormaterials/ParsedFunctorMaterial.C:127]:
```suggestion
  _func_params.resize(_n_functors + Moose::dim + 1);
```

File: test/tests/functormaterials/parsed_functor_material/tests [hunk 6]
```diff
@@ -0,0 +1,11 @@
+[Tests]
+  issues = '#24380'
+  design = 'ParsedFunctorMaterial.md'
+
```
Comment (lindsayad) [inline test/tests/functormaterials/parsed_functor_material/tests:4]:
```suggestion
```

File: test/tests/functormaterials/parsed_functor_material/parsed_functor_material.i [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+  zmax = 10.0
+[]
+
+[Functions]
+  [fn1]
+    type = ConstantFunction
+    value = 5
+  []
+  [fn2]
+    type = ConstantFunction
```
Comment (lindsayad) [inline test/tests/functormaterials/parsed_functor_material/parsed_functor_material.i:21]: Could we do something a little more exotic than a constant function, just to be triply sure that the space and time arguments are getting properly forwarded to the functors used by the parsed functor?

**Review by lindsayad — CHANGES_REQUESTED** [review]
This is going to be such an incredibly useful capability. 🤯

File: framework/src/functormaterials/ParsedFunctorMaterial.C (hunk 2 above)
Comment (joshuahansel) [reply framework/src/functormaterials/ParsedFunctorMaterial.C:85]: Works for me

**Review by lindsayad — COMMENTED** [review]
Dope. I'm good with the design

**Comment by joshuahansel** [comment]
Pronghorn will need a patch - I'll prepare that. Griffin build failures are unrelated.

**Review by GiudGiud — APPROVED** [review]
looks great

**Comment by joshuahansel** [comment]
Pronghorn fix: https://github.inl.gov/ncrc/pronghorn/pull/1220

**Review by lindsayad — APPROVED** [review]

**Pull request merged**